velox/functions/sparksql/Hash.cpp

```diff
@@ -26,21 +26,273 @@ namespace {
+    auto offset = base_->offsetAt(indices_[index]);
+
+    ReturnType result = seed;
+    for (auto i = 0; i < size; ++i) {
+      result = elementHasher_->hashAt(i + offset, result);
+    }
+    return result;
+  }
+
+ private:
+  const ArrayVector* base_;
+  const int32_t* indices_;
+  DecodedVector decodedElements_;
+  std::shared_ptr<SparkVectorHasher<HashClass>> elementHasher_;
+};
+
+template <typename HashClass, typename SeedType, typename ReturnType>
+class MapVectorHasher : public SparkVectorHasher<HashClass> {
+ public:
+  MapVectorHasher(DecodedVector& decoded)
+      : SparkVectorHasher<HashClass>(decoded) {
+    base_ = decoded.base()->as<MapVector>();
+    indices_ = decoded.indices();
+
+    SelectivityVector rows(base_->mapKeys()->size());
+    decodedKeys_.decode(*base_->mapKeys(), rows);
+    decodedValues_.decode(*base_->mapValues(), rows);
+    keyHasher_ = createVectorHasher<HashClass>(decodedKeys_);
+    valueHasher_ = createVectorHasher<HashClass>(decodedValues_);
+  }
+
+  ReturnType hashNotNullAt(vector_size_t index, SeedType seed) override {
+    auto size = base_->sizeAt(indices_[index]);
+    auto offset = base_->offsetAt(indices_[index]);
+
+    ReturnType result = seed;
+    for (auto i = 0; i < size; ++i) {
+      result = keyHasher_->hashAt(i + offset, result);
+      result = valueHasher_->hashAt(i + offset, result);
+    }
+    return result;
+  }
+
+ private:
+  const MapVector* base_;
+  const int32_t* indices_;
+  DecodedVector decodedKeys_;
+  DecodedVector decodedValues_;
+  std::shared_ptr<SparkVectorHasher<HashClass>> keyHasher_;
+  std::shared_ptr<SparkVectorHasher<HashClass>> valueHasher_;
+};
+
+template <typename HashClass, typename SeedType, typename ReturnType>
+class RowVectorHasher : public SparkVectorHasher<HashClass> {
+ public:
+  RowVectorHasher(DecodedVector& decoded)
+      : SparkVectorHasher<HashClass>(decoded) {
+    base_ = decoded.base()->as<RowVector>();
+    indices_ = decoded.indices();
+
+    SelectivityVector rows(base_->size());
+    decodedChildren_.resize(base_->childrenSize());
+    hashers_.resize(base_->childrenSize());
+    for (auto i = 0; i < base_->childrenSize(); ++i) {
+      decodedChildren_[i].decode(*base_->childAt(i), rows);
+      hashers_[i] = createVectorHasher<HashClass>(decodedChildren_[i]);
+    }
+  }
+
+  ReturnType hashNotNullAt(vector_size_t index, SeedType seed) override {
+    ReturnType result = seed;
+    for (auto i = 0; i < base_->childrenSize(); ++i) {
+      result = hashers_[i]->hashAt(indices_[index], result);
+    }
+    return result;
+  }
+
+ private:
+  const RowVector* base_;
+  const int32_t* indices_;
+  std::vector<DecodedVector> decodedChildren_;
+  std::vector<std::shared_ptr<SparkVectorHasher<HashClass>>> hashers_;
+};
+
 // ReturnType can be either int32_t or int64_t
 // HashClass contains the function like hashInt32
-template <typename ReturnType, typename HashClass, typename SeedType>
+template <
```

Comment:
do we need to carry these three template types through all the intermediate classes? Could we instead just pass `HashClass`, and use the trait class only when we actually need SeedType and ReturnType, which is in the hashOne() functions?